File: Eduflex/Eduflex/ShareService/DataAccess/Service/Department.cs
using MongoDB.Driver;
using ShareService.Common;
using ShareService.DataAccess.Common;
using ShareService.DataAccess.Interface;
using ShareService.Models.Department;

namespace ShareService.DataAccess
{
    public class Department : AuditableCollectionBase<DepartmentModel>, IDepartment
    {
        public Department(IMongoDatabase database, ICurrentUserService currentUser)
            : base(database.GetCollection<DepartmentModel>("Departments"), currentUser)
        {
        }

        public async Task<bool> CreateDepartmentAsync(DepartmentModel department)
        {
            await InsertOneAsync(department);
            return true;
        }

        public async Task<DepartmentModel?> GetDepartmentByIdAsync(string id)
        {
            return await Collection.Find(d => d.Id == id).FirstOrDefaultAsync();
        }

        public async Task<DepartmentModel?> GetByNameAsync(string name)
        {
            return await Collection.Find(d => d.Name == name).FirstOrDefaultAsync();
        }

        public async Task<List<DepartmentModel>> GetAllAsync()
        {
            return await Collection.Find(FilterDefinition<DepartmentModel>.Empty).ToListAsync();
        }

        public Task<PagedResult<DepartmentModel>> GetDepartmentsAsync(DepartmentFilter filter)
        {
            var mongoFilter = BuildSearchFilter(filter.SearchTerm, d => d.Name);
            var sort = Builders<DepartmentModel>.Sort.Ascending(d => d.Name);

            return GetPagedAsync(mongoFilter, sort, filter.PageNumber, filter.PageSize);
        }

        public async Task<bool> UpdateDepartmentAsync(string id, DepartmentModel department)
        {
            return await ReplaceOneAsync(d => d.Id == id, department);
        }

        public async Task<bool> DeleteDepartmentAsync(string id)
        {
            var result = await Collection.DeleteOneAsync(d => d.Id == id);
            return result.DeletedCount > 0;
        }

        public async Task<bool> HasChildDepartmentsAsync(string departmentId)
        {
            return await Collection.Find(d => d.ParentDepartmentId == departmentId).AnyAsync();
        }
    }
}

File: eduflex-Angular/src/app/models/department.ts
export interface Department {
  id: string;
  name: string;
  description?: string;
  parentDepartmentId?: string;
  headUserId?: string;
  memberUserIds: string[];
  createdAt: string;
}

export interface CreateDepartmentRequest {
  name: string;
  description?: string;
  parentDepartmentId?: string;
  headUserId?: string;
  memberUserIds: string[];
}

export interface DepartmentFilter {
  pageNumber: number;
  pageSize: number;
  searchTerm?: string;
}

export interface DepartmentBadge {
  id: string;
  name: string;
  isHead: boolean;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
}
